src/SecTester.Repeater/Bus/IncomingRequest.cs

```diff
@@ -1,19 +1,93 @@
 using System;
 using System.Collections.Generic;
+using System.IO;
+using System.Linq;
 using System.Net.Http;
+using System.Reflection;
+using System.Runtime.Serialization;
 using MessagePack;
-using SecTester.Core.Bus;
+using SecTester.Repeater.Bus.Formatters;
 using SecTester.Repeater.Runners;
 
 namespace SecTester.Repeater.Bus;
 
-[MessagePackObject(true)]
-public record IncomingRequest(Uri Url) : Event, IRequest
+[MessagePackObject]
+public record IncomingRequest(Uri Url) : HttpMessage, IRequest
 {
-  public string? Body { get; set; }
+  private const string UrlKey = "url";
+  private const string MethodKey = "method";
+
+  private static readonly Dictionary<string, Protocol> ProtocolEntries = typeof(Protocol)
+    .GetFields(BindingFlags.Public | BindingFlags.Static)
+    .Select(field => new
+    {
+      Value = (Protocol)field.GetValue(null),
+      StringValue = field.GetCustomAttribute<EnumMemberAttribute>()?.Value ?? field.Name
+    })
+    .ToDictionary(x => x.StringValue, x => x.Value);
+
+  [Key(MethodKey)]
+  [MessagePackFormatter(typeof(MessagePackHttpMethodFormatter))]
   public HttpMethod Method { get; set; } = HttpMethod.Get;
-  public Protocol Protocol { get; set; } = Protocol.Http;
+
+  [Key(UrlKey)]
   public Uri Url { get; set; } = Url ?? throw new ArgumentNullException(nameof(Url));
-  public IEnumerable<KeyValuePair<string, IEnumerable<string>>> Headers { get; set; } =
-    new List<KeyValuePair<string, IEnumerable<string>>>();
+
+  public static IncomingRequest FromDictionary(Dictionary<object, object> dictionary)
```

Comment:
Why do not map fields automatically like this:

```csharp
[Key(HeadersKey)]
[MessagePackFormatter(typeof(MessagePackHttpHeadersFormatter))]
private IEnumerable<KeyValuePair<string, object?>> _headers;

public IEnumerable<KeyValuePair<string, IEnumerable<string>>> Headers
{
    get => _headers.Select(kvp => kvp.Value switch
    {
        IEnumerable<string> strings => new KeyValuePair<string, IEnumerable<string>>(kvp.Key, strings),
        null => new KeyValuePair<string, IEnumerable<string>>(kvp.Key, Enumerable.Empty<string>()),
        _ => new KeyValuePair<string, IEnumerable<string>>(kvp.Key, new[] { kvp.Value.ToString() })
    });
    set => _headers = value.Select(kvp => new KeyValuePair<string, object?>(kvp.Key, kvp.Value?.ToArray()));
}
``` 

The same could be applied to the rest of the fields.

Reply:
Looks promising, but this is not going to work. The SocketIO `PackMessage` is not compatible to `*AllowPrivate` resolvers.